### PR TITLE
8300829: Make CtwRunner available as an independent tool

### PR DESCRIPTION
**Pull request by caojoshua**
1. ~~Create an independent jar file with CtwRunner as the main class to make it easier to run~~. Not needed anymore thanks to @navyxliu 's comments. We can use the original `ctw.jar` because it already has ctwrunner class.
2. Output the class files directly into the destination directory. Currently, CTW expects a `wb.jar`, but the jtreg tests that use CTWRunner has class files outside of a jar.
3. Introduce `sun.hotspot.tools.ctwrunner.ctw_extra_args` option to pass extra arguments to CTW. Arguments are comma separated because working with spaces in bash can be kind of awkward, but I'm open to changing this part.

### Motivation
CTWRunner is a wrapper around CTW that will continue compiling beyond failure. It can be useful for testing compilation with certain flags. For example, one could run

```
JAVA_OPTIONS="-Dsun.hotspot.tools.ctwrunner.ctw_extra_args=-XX:+StressLCM,-XX:+StressGCM" ./ctwrunner.sh modules:java.base
```

To test compiling the java.base module with `-XX:+StressLCM -XX:+StressGCM`. This is advantageous over uses CTW because we can see the full list of crashes for the entire module.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300829](https://bugs.openjdk.org/browse/JDK-8300829): Make CtwRunner available as an independent tool

### Reviewers
 * [Xin Liu](https://openjdk.org/census#xliu) (@navyxliu - Committer) ⚠️ Review applies to [35979be8](https://git.openjdk.org/jdk/pull/13344/files/35979be830ae2aac3c0d0e007d5517c044938533)
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13344/head:pull/13344` \
`$ git checkout pull/13344`

Update a local copy of the PR: \
`$ git checkout pull/13344` \
`$ git pull https://git.openjdk.org/jdk.git pull/13344/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13344`

View PR using the GUI difftool: \
`$ git pr show -t 13344`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13344.diff">https://git.openjdk.org/jdk/pull/13344.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13344#issuecomment-1496915264)

### CODE REVIEW

File: test/hotspot/jtreg/testlibrary/ctw/Makefile [hunk 1]
```diff
@@ -78,22 +81,32 @@ $(DST_DIR)/ctw.sh: $(DST_DIR)
...
+	chmod a+x $@
+
 $(DST_DIR)/ctw.jar: filelist $(DST_DIR)/wb.jar
 	@mkdir -p $(OUTPUT_DIR)
 	$(JAVAC) $(EXPORTS) -sourcepath $(SRC_DIR) -d $(OUTPUT_DIR) -cp $(DST_DIR)/wb.jar @filelist
-	$(JAR) --create --file=$@ --main-class $(MAIN_CLASS) -C $(OUTPUT_DIR) .
+	$(JAR) --create --file=$@ --main-class $(CTW_MAIN_CLASS) -C $(OUTPUT_DIR) .
+	@rm -rf $(OUTPUT_DIR)
+
+$(DST_DIR)/ctwrunner.jar: filelist $(DST_DIR)/wb.jar
```
Comment (navyxliu) [inline test/hotspot/jtreg/testlibrary/ctw/Makefile:94]: Theoretically, we can avoid from generating ctwrunner.jar. The classes are the same as 'ctw.jar' and the only different part is the entry point. 

to launch CTWRunner, you can do thing like this in ctwruner.sh
```
echo '$${JAVA_HOME}/bin/java $${JAVA_OPTIONS} -Dtest.jdk=$${JAVA_HOME} -cp ctw.jar $CTWRUNNER_MAIN_CLASS  $$@' > $@
```

it's up to you. if we generate ctwrunner.jar, we will have a "standalone" jar.

File: test/hotspot/jtreg/testlibrary/ctw/src/sun/hotspot/tools/ctw/CtwRunner.java [hunk 2]
```diff
@@ -52,6 +53,14 @@ public class CtwRunner {
     private static final Predicate<String> IS_CLASS_LINE = Pattern.compile(
             "^\\[\\d+\\]\\s*\\S+\\s*$").asPredicate();
 
+    /**
+     * Value of {@code -Dsun.hotspot.tools.ctwrunner.ctw_extra_args}. Extra
+     * comma-separated arguments to pass to CTW subprocesses.
+     */
+    public static final String CTW_EXTRA_ARGS
```
Comment (navyxliu) [inline test/hotspot/jtreg/testlibrary/ctw/src/sun/hotspot/tools/ctw/CtwRunner.java:60]: how about you use getProperty("sun.hotspot.tools.ctwrunner.ctw_extra_args", "").
by giving it an empty string as default value, you can take iout if (null != CTW_EXTRA_ARGS) below. 

btw, you may also need to update the year in the copyrights header.

File: test/hotspot/jtreg/testlibrary/ctw/Makefile (hunk 1 above)
Comment (caojoshua) [reply test/hotspot/jtreg/testlibrary/ctw/Makefile]: I removed ctwrunner.jar in newest commit.

I also realized I was javac'ing twice. We avoid this entirely by removing the extra jar.

File: test/hotspot/jtreg/testlibrary/ctw/src/sun/hotspot/tools/ctw/CtwRunner.java (hunk 2 above)
Comment (caojoshua) [reply test/hotspot/jtreg/testlibrary/ctw/src/sun/hotspot/tools/ctw/CtwRunner.java:60]: Yup. Made these changes.

**Review by navyxliu — APPROVED** [review]
LGTM. I am not a reviewer. need other reviewers to approve this.

File: test/hotspot/jtreg/testlibrary/ctw/src/sun/hotspot/tools/ctw/CtwRunner.java (hunk 2 above)
Comment (phohensee) [reply test/hotspot/jtreg/testlibrary/ctw/src/sun/hotspot/tools/ctw/CtwRunner.java:60]: Better to make CTW_EXTRA_ARGS private. If it's needed outside this class, add a public accessor. Don't think it's needed outside though.

Comment (caojoshua) [reply test/hotspot/jtreg/testlibrary/ctw/src/sun/hotspot/tools/ctw/CtwRunner.java]: Made it private. Not needed outside.

**Review by phohensee — APPROVED** [review]
Lgtm.

**Pull request closed without merging**